### PR TITLE
Update Rust to 1.24.1

### PR DESCRIPTION
**Pull request by negator**
Upgrades Rust to take advantage of newer features, namely incremental compilation:
https://github.com/rust-lang/rust/blob/master/RELEASES.md

### CODE REVIEW

**Comment by acmcarther** [comment]
@buchgr: Can you take a look at this CI failure? It looks like it wasn't able to run the CI authorization process as the list of repository collaborators is privileged. If you don't have cycles, I can take a look later today

---

I'm not sure that incremental compilation will work "out of the box" in Bazel. In particular, if it does, that mean we're leaking state somewhere (this is probable), as we'll need to explicitly capture the cached state somewhere known to Bazel.

From what I can tell, this is something that needs upstream thought leadership. Swift folks seem to be having similar issues: https://github.com/bazelbuild/rules_apple/issues/21

**Comment by acmcarther** [comment]
More notes:

Scala folks have an incremental compiler called "Zinc". This issue (https://github.com/bazelbuild/rules_scala/issues/328) is a starting point for their integration.

An older discussion here has a lot of details on why it doesn't work out of the box: https://groups.google.com/d/msg/bazel-discuss/3iUy5jxS3S0/1Fd_IsIEBgAJ

**Comment by buchgr** [comment]
@acmcarther thanks for notifying. There was a permission error. It should be fixed now.

File: rust/repositories.bzl [hunk 1]
```diff
@@ -125,16 +125,16 @@ def rust_repositories():
   native.new_http_archive(
       name = "rust_linux_x86_64",
       url = "https://static.rust-lang.org/dist/rust-1.20.0-x86_64-unknown-linux-gnu.tar.gz",
```
Comment (acmcarther) [inline rust/repositories.bzl]: Forgot to change url (here and below)?

**Comment by acmcarther** [comment]
Thanks for the speedy fix buchgr@!

File: rust/repositories.bzl (hunk 1 above)
Comment (negator) [reply rust/repositories.bzl]: Oof sorry about that!

**Comment by acmcarther** [comment]
@negator: Belatedly, thanks for your contribution! I guess we missed three whole minor versions up until now....

I don't see any reason to delay submitting this, so I'll go ahead

**Comment by negator** [comment]
Excellent! Thanks for extremely quick responses!

**Comment by mfarrugi** [comment]
@negator it is also straight forward to manage versions outside of the rules, see #64 

@acmcarther We may want to make it easier to parameterize toolchain configuration, although not immediately clear how to remove the boilerplate in [this](https://github.com/mfarrugi/cargo-raze-example-stdx/blob/master/bazel/custom_rust_toolchain.bzl)

**Comment by negator** [comment]
@mfarrugi Yes thank you I did see that, and that was the approach I was initially considering. However, upgrading the libs seemed very straightforward. 

Additionally I was hoping to get incremental compilation for free, however as @acmcarther points out, this is not in fact for free. A persistent worker that invokes `rustc -C...` with a `tmp` dir appears to be the correct way forward. This also appears straightforward, but will require a fair amount of boilerplate. Rules closure could perhaps be used as a reference: https://github.com/bazelbuild/rules_closure/blob/master/java/io/bazel/rules/closure/ClosureWorker.java

**Comment by lilianmoraru** [comment]
Rust 1.25 is out.

**Pull request merged**